### PR TITLE
frontend: Home: Fix so it does not rerender a lot

### PR DESCRIPTION
**Pull request by illume**
Some of the hooks were causing a rerender several times every 5 seconds or so. Even when things have not changed. Some of the types have been added to make it more clear what data is being passed around.

Also now there is a HomePure presentational component.

Event.useWarningList and useClustersVersion needed a deep equality comparison to prevent rerendering properly.

### how to test

- [x] You can either use dev tools or add a console.log into the HomePure component to see how often it is rendered. With this change it should only rerender when something changes.
- [x] stop a cluster, and start a cluster... the home page should update to show the correct cluster active status

### CODE REVIEW

File: frontend/src/components/App/Home/index.tsx [hunk 1]
```diff
@@ -160,30 +161,106 @@ function ClusterStatus({ error }: { error?: ApiError | null }) {
...
+  const [versions, errors] = useDeepCompareMemo(versionsErrors, [versionsErrors]);
+  const warningsList = Event.useWarningList(customClusterNames);
+  const warningsMap = useDeepCompareMemo(warningsList, [warningsList]);
+
+  console.log('versions', versions);
+  console.log('clusterList', clusterList);
+  console.log('warningsList', warningsList);
+  console.log('customClusterNames.join', customClusterNames.join('-'));
+
+  const rendered = useMemo(() => {
```
Comment (sniok) [inline frontend/src/components/App/Home/index.tsx:230]: Please use React.memo with _.isEqual for the pure Home component, then most of the useMemos won't be necessary

File: frontend/src/components/App/Home/index.tsx [hunk 2]
```diff
@@ -160,30 +161,106 @@ function ClusterStatus({ error }: { error?: ApiError | null }) {
 
 export default function Home() {
   const history = useHistory();
-  const clusters = useClustersConf() || {};
+  const clustersConf = useClustersConf() || {};
+  const clusters = useMemo(() => {
+    return clustersConf;
+  }, [clustersConf]);
```
Comment (joaquimrocha) [inline frontend/src/components/App/Home/index.tsx:167]: I thought memo didn't compare objects. If that's the case, this should not do much.

File: frontend/src/components/App/Home/index.tsx [hunk 3]
```diff
@@ -160,30 +161,106 @@ function ClusterStatus({ error }: { error?: ApiError | null }) {
...
 
   if (!helpers.isElectron() && Object.keys(clusters).length === 1) {
     history.push(createRouteURL('cluster', { cluster: Object.keys(clusters)[0] }));
     return null;
   }
-
```
Comment (joaquimrocha) [inline frontend/src/components/App/Home/index.tsx:169]: Not needed.

File: frontend/src/components/App/Home/index.tsx [hunk 4]
```diff
@@ -160,30 +161,106 @@ function ClusterStatus({ error }: { error?: ApiError | null }) {
...
+  }, [customNameClusters]);
+
+  const clusterList = useMemo(() => {
+    return Object.values(clusters);
+  }, [clusters]);
+
+  const versionsErrors = useClustersVersion(clusterList);
+  const [versions, errors] = useDeepCompareMemo(versionsErrors, [versionsErrors]);
+  const warningsList = Event.useWarningList(customClusterNames);
+  const warningsMap = useDeepCompareMemo(warningsList, [warningsList]);
```
Comment (joaquimrocha) [inline frontend/src/components/App/Home/index.tsx:223]: Assuming the warnings are only updated when the actual warning objects in them change, I don't think this is making a lot of difference.

File: frontend/src/components/App/Home/index.tsx [hunk 5]
```diff
@@ -160,30 +161,106 @@ function ClusterStatus({ error }: { error?: ApiError | null }) {
...
+    return (
+      <HomePure
+        customNameClusters={customNameClusters}
+        customClusterNames={customClusterNames}
+        versions={versions}
+        errors={errors}
+        warningsMap={warningsMap}
+      />
+    );
+  }, [customClusterNames, customClusterNames, versions, errors, warningsMap]);
```
Comment (joaquimrocha) [inline frontend/src/components/App/Home/index.tsx:240]: I don't understand how the memo helps here. The component shouldn't refresh anyway unless one of its state or props changes, and AFAICS, the properties are now all extracted and being passed to them but also added here as deps, so this is having the same effect as if we just pass the properties.

**Review by joaquimrocha — CHANGES_REQUESTED** [review]
I think a much more effective and simpler change may be to fix useWarningList and useClustersVersion so it doesn't update unless something in the end result is changed.

**Pull request closed without merging**